src/id_allocator.rs

```diff
@@ -0,0 +1,182 @@
+// Copyright 2022 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// Copyright © 2019 Intel Corporation. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0 OR BSD-3-Clause
+
+//! Provides allocation and releasing strategy for IDs.
+//!
+//! This module implements an allocation strategies for all resource types
+//! that can be abstracted to an integer.
+
+use crate::Error;
+use crate::Result;
+use std::collections::BTreeSet;
+
+/// Internal representation of IdAllocator. Contains the ends of the interval
+/// that is managed, a field that points at the next available ID, and a
+/// BTreeSet used to store the released IDs. The reason we have chosed a
+/// BTreeSet is that the average complexity for deletion and insertion is
+/// O(logN) compared to Vec for example, another benefit is that the entries
+/// are sorted so we will always use the first available ID.
+#[derive(Debug)]
+pub struct IdAllocator {
+    // Beginnbing of the range of IDs that we want to manage.
+    range_base: u32,
+    // First available id that was never allocated.
+    next_id: Option<u32>,
+    // End of the range of IDs that we want to manage.
+    range_end: u32,
+    // Set of all freed ids that can be reused at subsequent allocations.
+    freed_ids: BTreeSet<u32>,
+}
+
+impl IdAllocator {
+    /// Creates a new instance of IdAllocator that will be used to manage the
+    /// allocation and release of ids from the interval specified by
+    /// `range_base` and `range_max`
+    pub fn new(range_base: u32, range_end: u32) -> std::result::Result<Self, Error> {
+        if range_end <= range_base {
```

Comment:
I think the check should be `range_end < range_base`. the `allocate_id()` code suggests that the range is inclusive, therefore `IdAllocator::new(1,1);` should be allowed, because it contains a range with one element.